### PR TITLE
Riscv overlay software standard v2

### PR DESCRIPTION
**Pull request by oferShinaar**

### CODE REVIEW

File: riscv-overlay-software-standard-draft.adoc [hunk 1]
```diff
@@ -263,44 +304,97 @@ The descriptor will be materialized at link time.
...
-d.  This table shall provide sufficient information for the RT-Engine, debugger or other utilities to find the requested group within the *_"overlay area"_* (for example, so that the FW can find the locate and load a group).
-9.  Overlay group size ranges from 512B – 4K for both functions and data.
+.. An overlay function can be resident in more than one group.
+.. Overlay data can be resident in more than one group.
 
+.  Overlay Static Table
+..  The linker shall create a group-offset-table to hold all the overlay group offsets. Each entry index in the table represents an overlay group ID. Each entry contents represent the zero base offset to the group.
+..  Overlay group IDs are numerical.
+..  This table can be read at runtime (e.g. by the RT-Engine, debugger or another utility) to provide a mapping to locate an overlay group.
+..  This table shall provide sufficient information for the RT-Engine, debugger or other utilities to find the requested group within the *_"overlay area"_* (for example, so that the FW can find the locate and load a group).
```
Comment (craigblackmore) [inline riscv-overlay-software-standard-draft.adoc:323]: `find the locate` -> `locate`

File: riscv-overlay-software-standard-draft.adoc [hunk 2]
```diff
@@ -263,44 +304,97 @@ The descriptor will be materialized at link time.
...
 
+===== Linker flags
+The linker will get all the necessary data for overlay symbols from: object files, the linker script and linker flags.
+
+... *Input file*
++
+An external file holds "group numbers" per function name
+(this is for manual grouping). Appendix to file format  _(link:#Appendix-A-Grouping-file-syntax[Appendix A - Grouping file syntax])_.
+Without providing this file, the linker will generate a group per function
+... *Max / Min size of overaly group*
```
Comment (craigblackmore) [inline riscv-overlay-software-standard-draft.adoc:336]: `overaly` -> `overlay`

File: riscv-overlay-software-standard-draft.adoc [hunk 3]
```diff
@@ -263,44 +304,97 @@ The descriptor will be materialized at link time.
...
++
+For the linker to be aware of the selected group size, the user should reflect
```
Comment (craigblackmore) [inline riscv-overlay-software-standard-draft.adoc:338]: `reflect` -> `specify`?

File: riscv-overlay-software-standard-draft.adoc [hunk 4]
```diff
@@ -263,44 +304,97 @@ The descriptor will be materialized at link time.
...
+
+
+[[Appendix-A-Grouping-file-syntax]]
+= Appendix A - Grouping file syntax
+
+The linker can receive an input file to give it details about assignments of groups to functions.
+e.g. myFunction() should be in group 1.
+This file should be in comma-separate-value syntax (CSV), as described:
+
+*	Each new line represent a function
```
Comment (craigblackmore) [inline riscv-overlay-software-standard-draft.adoc:383]: `represent` -> `represents`

File: riscv-overlay-software-standard-draft.adoc [hunk 5]
```diff
@@ -263,44 +304,97 @@ The descriptor will be materialized at link time.
...
+*	First column holds a function name
+*	Each next column holds a group number to assign the function
+
+[[Example-for-Grouping-file-CSV]]
+.Example for Grouping file CSV
+
+[%header, format=csv]
+|===
+*Function name* , *Group number*, *Group number*, *Group ...*
```
Comment (craigblackmore) [inline riscv-overlay-software-standard-draft.adoc:392]: Will the header be required or optional?

File: riscv-overlay-software-standard-draft.adoc [hunk 6]
```diff
@@ -131,39 +140,48 @@ A ‘Group’ is a collection of overlay functions and overlay data. We should u
...
 
 The RT-Engine will be the entity to invoke the overlay function.
 
 1.  The RT-Engine should support invoking indirect function calls (i.e. calls via function-pointers)
-2.  The search-algorithm is open to interpretation; we recommend to have at least one, for example LRU.
 3.  After a function is loaded to the heap, the RT-Engine will be responsible for passing all requested arguments from the root caller to the callee.
 +
 Therefore, the RT-Engine will apply the ABI rules.
-4.  Following the RISCV psABI, we should support #8 argument registers.
+4.  We should follow the ABI in the matter of passing X numbers of arguments to callee. Per the current RISCV psABI the max numbers of registers to pass to callee is #8.
```
Comment (craigblackmore) [inline riscv-overlay-software-standard-draft.adoc:183]: Return values should also be passed back to the caller in accordance with the ABI.

**Review by craigblackmore — COMMENTED** [review]
This looks ok. I have added a few minor comments.

File: riscv-overlay-software-standard-draft.adoc (hunk 5 above)
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc:392]: Good question. let's have a header, and on the implementation side, we will check if there is a header or not (to support both cases)

File: riscv-overlay-software-standard-draft.adoc (hunk 6 above)
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc:183]: to add the following ?
5. Return values to caller should also respect the ABI

File: riscv-overlay-software-standard-draft.adoc [hunk 7]
```diff
@@ -250,10 +275,26 @@ The main compiler demands are related to generating a sequence code to enter the
 3.  Types of related overlay use cases:
 a.  Direct call – just calling to the overlay function
 b.  Indirect call – call is via function pointer
-c.  Data – data which is marked as overlay should be referenced with the same sequence to enter the RT-Engine so that the data can be loaded as necessary.
-4.  We probably need to reserve a few core registers to be used only for the RT-Engine. Those registers will have a special purpose understood by the compiler, linker, debugger and RT-Engine. Moreover those registers form a _RT-Eng-Debugger-handshake_ between compiler, RT code and debugger.
+c.  RO Data – read-only data which is marked as overlay should be referenced with the same sequence to enter the RT-Engine so that the data can be loaded as necessary.
+4.  We need to reserve N core registers. To be used only for the RT-Engine.
+Those registers will have a special purpose understood by the compiler, linker, debugger, and RT-Engine.
+Moreover, those registers form a _RT-Eng-Debugger-handshake_ between compiler, RT code, and debugger.
+Any library linked with an overlay enabled it to be compiled without using the reserved registers.
```
[the remark this answers is not in the document]
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc:282]: fixed in #ca5d396

File: riscv-overlay-software-standard-draft.adoc [hunk 8]
```diff
@@ -131,39 +140,48 @@ A ‘Group’ is a collection of overlay functions and overlay data. We should u
...
 1.  The heap area should be defined pre-build.
-2.  The heap should have range limitation to be in sync with the RT-Engine and toolchain. The heap minimum size should not be smaller than the maximum pre-defined overlay group size.
+2.  The heap should have range limitation to be in sync with the RT-Engine and toolchain. The heap minimum size should be bigger or equal to the maximum pre-defined overlay group size.
+(*_heap-min-size >= max group size_*)
 3.  We can have multiple heaps to be controlled by a single/multiple RT-Engine(s).
 4.  A _Load-Function-Hook_ footnote:[Hook implementation will be the responsibility of the platform since only the platform knows how to implement them. +
 Please refer to section *_3.1.6 Platform/Framework Hooks_*] will be provided to the user for executing the load operation itself.
+5. The RT-Engine should hold all information regarding the heap. Which area is allocated/free, sizes, and address.
+6. Based on the given "heap information," the RT-Engine should provide a mechanism to lock segments in the heaps.
+7. The heap information should be acceptable to the application to get the status of the heap.
```
[the remark this answers is not in the document]
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc:169]: fixed in #ca5d396

File: riscv-overlay-software-standard-draft.adoc (hunk 1 above)
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc:323]: fixed in #ca5d396

File: riscv-overlay-software-standard-draft.adoc (hunk 2 above)
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc:336]: fixed in #ca5d396

File: riscv-overlay-software-standard-draft.adoc (hunk 3 above)
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc:338]: fixed in #ca5d396

File: riscv-overlay-software-standard-draft.adoc (hunk 4 above)
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc:383]: fixed in #ca5d396

File: riscv-overlay-software-standard-draft.adoc [hunk 9]
```diff
@@ -131,39 +140,48 @@ A ‘Group’ is a collection of overlay functions and overlay data. We should u
...
 1.  The heap area should be defined pre-build.
-2.  The heap should have range limitation to be in sync with the RT-Engine and toolchain. The heap minimum size should not be smaller than the maximum pre-defined overlay group size.
+2.  The heap should have range limitation to be in sync with the RT-Engine and toolchain. The heap minimum size should be bigger or equal to the maximum pre-defined overlay group size.
+(*_heap-min-size >= max group size_*)
 3.  We can have multiple heaps to be controlled by a single/multiple RT-Engine(s).
 4.  A _Load-Function-Hook_ footnote:[Hook implementation will be the responsibility of the platform since only the platform knows how to implement them. +
 Please refer to section *_3.1.6 Platform/Framework Hooks_*] will be provided to the user for executing the load operation itself.
+5. The RT-Engine should hold all information regarding the heap. Which area is allocated/free, sizes, and address.
+6. Based on the given "heap information," the RT-Engine should provide a mechanism to lock segments in the heaps.
+7. The heap information should be acceptable by the application to get the status of the heap.
```
Comment (craigblackmore) [inline riscv-overlay-software-standard-draft.adoc:169]: `acceptable` -> `accessible`

File: riscv-overlay-software-standard-draft.adoc [hunk 10]
```diff
@@ -279,7 +279,7 @@ c.  RO Data – read-only data which is marked as overlay should be referenced w
 4.  We need to reserve N core registers. To be used only for the RT-Engine.
 Those registers will have a special purpose understood by the compiler, linker, debugger, and RT-Engine.
 Moreover, those registers form a _RT-Eng-Debugger-handshake_ between compiler, RT code, and debugger.
-Any library linked with an overlay enabled it to be compiled without using the reserved registers.
+Any library linked with an overlay enabled application compiled without using the reserved registers.
```
Comment (craigblackmore) [inline riscv-overlay-software-standard-draft.adoc:282]: `application compiled` -> `application must be compiled`

File: riscv-overlay-software-standard-draft.adoc (hunk 6 above)
Comment (craigblackmore) [reply riscv-overlay-software-standard-draft.adoc:183]: Yes

File: riscv-overlay-software-standard-draft.adoc (hunk 10 above)
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc:282]: I think we lost context here. The meaning of this sentence is to say that: if you have libs linked to your application, which use overlay, you must compile those libs again just without the "reserve regs"
I fixed it in: #9c35a854a4a8e789db3f6078b0bc88cc47605807

File: riscv-overlay-software-standard-draft.adoc (hunk 9 above)
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc:169]: fixed in: #9c35a854a4a8e789db3f6078b0bc88cc47605807

File: riscv-overlay-software-standard-draft.adoc (hunk 6 above)
Comment (oferShinaar) [reply riscv-overlay-software-standard-draft.adoc]: fixed in: #9c35a854a4a8e789db3f6078b0bc88cc47605807

**Review by craigblackmore — APPROVED** [review]

**Pull request merged**